### PR TITLE
Increase MAX_D1D, MAX_Q1D, etc. when target is CPU

### PR DESCRIPTION
**Pull request by pazner**
This PR makes the way we handle `MAX_D1D`, `MAX_Q1D`, etc. more systematic and flexible. In particular, it allows for different values on different architectures (CUDA, HIP, CPU, etc.), so that a user running on CPU isn't bound by the same limitations as a user running on GPU. Specifically, the CUDA limitations should only take effect if the user selects the CUDA device at _runtime_, but not if they compiled with CUDA enabled, but are running on the CPU.

The main structure is in `forall.hpp`. There are two APIs to use the limit values:

1. In host code, when checking that the limits are satisfied (e.g. with `MFEM_VERIFY` or similar), the user can get the values from the return value of `DeviceDofQuadLimits::Get()`. This will return the limits of the configured `Device`, e.g. according to whether `-d cuda`, `-d hip`, or `-d cpu` is specified at the command line. Since these depend on runtime parameters, the results are not `constexpr` and cannot be used as template parameters, etc.
2. In kernels or host-device functions (e.g. `mfem::forall` kernels or functions defined with `MFEM_HOST_DEVICE`), the type alias `DofQuadLimits` will resolve to either `DofQuadLimits_CUDA`, `DofQuadLimits_HIP`, or `DofQuadLimits_CPU`, depending on whether the function is being compiled for the CPU or the GPU (i.e. depending on if `__CUDA_ARCH__` or `__HIP_DEVICE_COMPILE__` are defined). This provides access to the limits as `static constexpr` variables, which can be used for statically sized arrays, etc.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3722](https://github.com/mfem/mfem/pull/3722) | @pazner | @tzanio | @v-dobrev + @tzanio + @camierjs | 7/1/23 | 9/25/23 | 9/26/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
One suggestion: it is better to switch the max values based on the compilation pass, host or device -- this way even if compiling with CUDA, the host code will use the bigger value. In CUDA, the macro to check is `__CUDA_ARCH__` (compiling for device if defined, otherwise compiling for host). For HIP, the respective macro is `__HIP_DEVICE_COMPILE__`. We can define a single MFEM-specific macro (e.g. `MFEM_DEVICE_COMPILE`) to distinguish the host and device compilation passes.

**Comment by pazner** [comment]
> One suggestion: it is better to switch the max values based on the compilation pass, host or device -- this way even if compiling with CUDA, the host code will use the bigger value. In CUDA, the macro to check is `__CUDA_ARCH__` (compiling for device if defined, otherwise compiling for host). For HIP, the respective macro is `__HIP_DEVICE_COMPILE__`. We can define a single MFEM-specific macro (e.g. `MFEM_DEVICE_COMPILE`) to distinguish the host and device compilation passes.

That is a good idea, but that would mean that the checks would have to be moved inside the kernels themselves, correct? I think right now many of the `MFEM_VERIFY` checks are in host code, even when the kernel will be executed on device…

**Review by tzanio — APPROVED** [review]
Thanks for adding this, @pazner !

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
@camierjs and @v-dobrev, can you please review this when you get a chance?

**Comment by tzanio** [comment]
@pazner -- can you resolve conflicts please?

**Comment by pazner** [comment]
> @pazner -- can you resolve conflicts please?

Done in https://github.com/mfem/mfem/pull/3722/commits/15107fac437e9bb21f1d95df7b4d42013adc10e6

File: fem/integ/bilininteg_convection_ea.cpp [hunk 1]
```diff
@@ -157,8 +157,8 @@ static void EAConvectionAssemble3D(const int NE,
 {
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_D1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_Q1D, "");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_convection_ea.cpp:161]: `Q1D` was replaced with `D1D`-- I assume unintentionally.

File: fem/integ/bilininteg_convection_pa.cpp [hunk 2]
```diff
@@ -203,8 +203,8 @@ void PAConvectionApply2D(const int ne,
    const int NE = ne;
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_D1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_Q1D, "");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_convection_pa.cpp:207]: `Q1D` was replaced with `D1D`-- I assume unintentionally.

File: fem/integ/bilininteg_convection_pa.cpp [hunk 3]
```diff
@@ -323,8 +323,8 @@ void SmemPAConvectionApply2D(const int ne,
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
    constexpr int NBZ = T_NBZ ? T_NBZ : 1;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_D1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_Q1D, "");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_convection_pa.cpp:327]: `Q1D` was replaced with `D1D`-- I assume unintentionally.

File: fem/integ/bilininteg_convection_pa.cpp [hunk 4]
```diff
@@ -450,8 +450,8 @@ void PAConvectionApply3D(const int ne,
    const int NE = ne;
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_D1D, "");
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().MAX_Q1D, "");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_convection_pa.cpp:454]: Same here. I'll stop adding this comment -- just check for other places where this change was made.

File: fem/integ/bilininteg_hcurl_kernels.hpp [hunk 5]
```diff
@@ -59,8 +59,10 @@ inline void SmemPAHcurlMassAssembleDiagonal3D(const int d1d,
                                               const Vector &pa_data,
                                               Vector &diag)
 {
-   MFEM_VERIFY(T_D1D || d1d <= HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
-   MFEM_VERIFY(T_Q1D || q1d <= HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
+   MFEM_VERIFY(T_D1D ||
+               d1d <= DeviceDofQuadLimits::Get().HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
+   MFEM_VERIFY(T_Q1D ||
+               q1d <= DeviceDofQuadLimits::Get().HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurl_kernels.hpp:65]: The newline will look better just before the second argument of the macro.

File: fem/integ/bilininteg_hcurl_kernels.hpp [hunk 6]
```diff
@@ -218,8 +220,10 @@ inline void SmemPAHcurlMassApply3D(const int d1d,
                                    const Vector &x,
                                    Vector &y)
 {
-   MFEM_VERIFY(T_D1D || d1d <= HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
-   MFEM_VERIFY(T_Q1D || q1d <= HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
+   MFEM_VERIFY(T_D1D ||
+               d1d <= DeviceDofQuadLimits::Get().HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
+   MFEM_VERIFY(T_Q1D ||
+               q1d <= DeviceDofQuadLimits::Get().HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurl_kernels.hpp:226]: Same here -- move the newline before the second arg.

File: fem/integ/bilininteg_hcurl_kernels.hpp [hunk 7]
```diff
@@ -441,8 +445,10 @@ inline void PACurlCurlAssembleDiagonal3D(const int d1d,
                                          const Vector &pa_data,
                                          Vector &diag)
 {
-   MFEM_VERIFY(T_D1D || d1d <= HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
-   MFEM_VERIFY(T_Q1D || q1d <= HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
+   MFEM_VERIFY(T_D1D ||
+               d1d <= DeviceDofQuadLimits::Get().HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
+   MFEM_VERIFY(T_Q1D ||
+               q1d <= DeviceDofQuadLimits::Get().HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurl_kernels.hpp:451]: Same here.

File: fem/integ/bilininteg_hcurl_kernels.hpp [hunk 8]
```diff
@@ -635,8 +641,10 @@ inline void SmemPACurlCurlAssembleDiagonal3D(const int d1d,
                                              const Vector &pa_data,
                                              Vector &diag)
 {
-   MFEM_VERIFY(T_D1D || d1d <= HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
-   MFEM_VERIFY(T_Q1D || q1d <= HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
+   MFEM_VERIFY(T_D1D ||
+               d1d <= DeviceDofQuadLimits::Get().HCURL_MAX_D1D, "Error: d1d > HCURL_MAX_D1D");
+   MFEM_VERIFY(T_Q1D ||
+               q1d <= DeviceDofQuadLimits::Get().HCURL_MAX_Q1D, "Error: q1d > HCURL_MAX_Q1D");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_hcurl_kernels.hpp:647]: Same here. I'll stop posting about these -- just check for the same bad formatting below.

File: fem/integ/bilininteg_convection_ea.cpp [hunk 9]
```diff
@@ -38,7 +38,7 @@ static void EAConvectionAssemble1D(const int NE,
    {
       const int D1D = T_D1D ? T_D1D : d1d;
       const int Q1D = T_Q1D ? T_Q1D : q1d;
-      constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : DofQuadLimits::MAX_D1D;
```
Comment (v-dobrev) [inline fem/integ/bilininteg_convection_ea.cpp:41]: Should this be `MAX_Q1D`? Looks like `MAX_Q1D` was replaced by `MAX_D1D`.

File: fem/integ/bilininteg_mass_kernels.cpp [hunk 10]
```diff
@@ -25,16 +25,13 @@ static void PAMassAssembleDiagonal1D(const int NE,
                                      const int D1D,
                                      const int Q1D)
 {
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
```
Comment (v-dobrev) [inline fem/integ/bilininteg_mass_kernels.cpp:29]: Shouldn't we have a check here? This code was just removed, not replaced.

File: fem/integ/bilininteg_mass_kernels.cpp [hunk 11]
```diff
@@ -25,16 +25,13 @@ static void PAMassAssembleDiagonal1D(const int NE,
...
    auto B = Reshape(b.Read(), Q1D, D1D);
    auto D = Reshape(d.Read(), Q1D, NE);
    auto Y = Reshape(y.ReadWrite(), D1D, NE);
    mfem::forall(NE, [=] MFEM_HOST_DEVICE (int e)
    {
       for (int dx = 0; dx < D1D; ++dx)
       {
-         Y(dx, e) = 0.0;
```
Comment (v-dobrev) [inline fem/integ/bilininteg_mass_kernels.cpp:37]: This is bugfix, right? This method should perform `+=`, not `=`.

File: fem/integ/bilininteg_mass_kernels.cpp (hunk 10 above)
Comment (v-dobrev) [reply fem/integ/bilininteg_mass_kernels.cpp:29]: Maybe this is fine -- there are no arrays here and sizes are dynamic.

File: fem/qinterp/det.cpp [hunk 12]
```diff
@@ -133,12 +128,24 @@ static void Det3D(const int NE,
    double *GM = nullptr;
    if (!SMEM)
    {
-      d_buff->SetSize(2*MSZ*GRID);
+      const DeviceDofQuadLimits &limits = DeviceDofQuadLimits::Get();
+      const int max_q1d = T_Q1D ? T_Q1D : (SMEM ? limits.MAX_DET_1D : limits.MAX_D1D);
+      const int max_d1d = T_D1D ? T_D1D : (SMEM ? limits.MAX_DET_1D : limits.MAX_Q1D);
```
Comment (v-dobrev) [inline fem/qinterp/det.cpp:133]: `SMEM` is `false` here, so the expressions can be simplified a bit.

File: fem/qinterp/det.cpp [hunk 13]
```diff
@@ -75,8 +75,8 @@ static void Det2D(const int NE,
 
    mfem::forall_2D_batch(NE, Q1D, Q1D, NBZ, [=] MFEM_HOST_DEVICE (int e)
    {
-      constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
-      constexpr int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : DofQuadLimits::MAX_Q1D;
+      constexpr int MD1 = T_D1D ? T_D1D : DofQuadLimits::MAX_D1D;
```
Comment (v-dobrev) [inline fem/qinterp/det.cpp:79]: The template parameters `MAX_D1D` and `MAX_Q1D` are not used anymore, so they can be removed.

File: fem/integ/bilininteg_mass_kernels.cpp (hunk 10 above)
Comment (pazner) [reply fem/integ/bilininteg_mass_kernels.cpp:29]: That's right, the checks were unnecessary.

File: fem/integ/bilininteg_mass_kernels.cpp (hunk 11 above)
Comment (pazner) [reply fem/integ/bilininteg_mass_kernels.cpp:37]: Yes, this was a bug fix from 04d34e4de159423bff299e53a57772d725e05011

File: fem/qinterp/det.cpp (hunk 13 above)
Comment (pazner) [reply fem/qinterp/det.cpp:79]: Fixed in [c80186c](https://github.com/mfem/mfem/pull/3722/commits/c80186cecaa1cd3fe4a288d239cf119be85703b5)

File: fem/qinterp/det.cpp (hunk 12 above)
Comment (pazner) [reply fem/qinterp/det.cpp]: Fixed in [b89295d](https://github.com/mfem/mfem/pull/3722/commits/b89295dec791764fc5d4890c1cac8692b2025fe0)

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
It's may be good to mention these updates in the `CHANGELOG`.

**Comment by tzanio** [comment]
Can we mention this in `CHANGELOG` please?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**